### ISSUE TITLE
[ISNet] `MaxPool` padding values do not match with ONNX and TF in some patterns

### ISSUE THREAD

**Issue by PINTO0309** [PINTO0309/onnx2tf#207]
### Issue Type

Others

### onnx2tf version number

1.7.7

### onnx version number

1.13.1

### tensorflow version number

1.12.0rc0

### Download URL for ONNX

https://s3.ap-northeast-2.wasabisys.com/temp-models/onnx2tf_207/isnet_general_use_180x320.onnx.zip
https://s3.ap-northeast-2.wasabisys.com/temp-models/onnx2tf_207/isnet_general_use_480x640.zip

### Parameter Replacement JSON

```json
N/A
```

### Description

1. Personal
2. `MaxPool` padding values do not match with ONNX and TF in some patterns. In the example below, we expect the output shape to be `[1, 32, 8, 10]` at `/stage1/pool5/MaxPool_output_0`, but it is actually `[1, 32, 7, 10]`. Therefore, it becomes `Skipped (Deleted or Shape Unmatched)` and an accuracy error occurs.

    This is a rather vexing problem.

    ```bash
    onnx2tf -i isnet_general_use_480x640.onnx
    ```
    ```
    ONNX and TF output value validation started =========================================
    INFO: validation_conditions: np.allclose(onnx_outputs, tf_outputs, rtol=0.0, atol=0.0001, equal_nan=True)
    INFO: onnx_output_name: /conv_in/Conv_output_0 tf_output_name: tf.math.add/Add:0 shape: (1, 64, 240, 320) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconvin/conv_s1/Conv_output_0 tf_output_name: tf.math.add_2/Add:0 shape: (1, 64, 240, 320) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconvin/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu/Relu:0 shape: (1, 64, 240, 320) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv1/conv_s1/Conv_output_0 tf_output_name: tf.math.add_4/Add:0 shape: (1, 32, 240, 320) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv1/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_1/Relu:0 shape: (1, 32, 240, 320) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/pool1/MaxPool_output_0 tf_output_name: tf.nn.max_pool2d/MaxPool2d:0 shape: (1, 32, 120, 160) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv2/conv_s1/Conv_output_0 tf_output_name: tf.math.add_6/Add:0 shape: (1, 32, 120, 160) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv2/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_2/Relu:0 shape: (1, 32, 120, 160) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/pool2/MaxPool_output_0 tf_output_name: tf.nn.max_pool2d_1/MaxPool2d:0 shape: (1, 32, 60, 80) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv3/conv_s1/Conv_output_0 tf_output_name: tf.math.add_8/Add:0 shape: (1, 32, 60, 80) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv3/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_3/Relu:0 shape: (1, 32, 60, 80) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/pool3/MaxPool_output_0 tf_output_name: tf.nn.max_pool2d_2/MaxPool2d:0 shape: (1, 32, 30, 40) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv4/conv_s1/Conv_output_0 tf_output_name: tf.math.add_10/Add:0 shape: (1, 32, 30, 40) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv4/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_4/Relu:0 shape: (1, 32, 30, 40) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/pool4/MaxPool_output_0 tf_output_name: tf.nn.max_pool2d_3/MaxPool2d:0 shape: (1, 32, 15, 20) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv5/conv_s1/Conv_output_0 tf_output_name: tf.math.add_12/Add:0 shape: (1, 32, 15, 20) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/rebnconv5/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_5/Relu:0 shape: (1, 32, 15, 20) dtype: float32 validate_result:  Matches 
    INFO: onnx_output_name: /stage1/pool5/MaxPool_output_0 tf_output_name: tf.nn.max_pool2d_4/MaxPool2d:0 shape: (1, 32, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/rebnconv6/conv_s1/Conv_output_0 tf_output_name: tf.math.add_14/Add:0 shape: (1, 32, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/rebnconv6/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_6/Relu:0 shape: (1, 32, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/rebnconv7/conv_s1/Conv_output_0 tf_output_name: tf.math.add_16/Add:0 shape: (1, 32, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/rebnconv7/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_7/Relu:0 shape: (1, 32, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/Concat_output_0 tf_output_name: tf.concat/concat:0 shape: (1, 64, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/rebnconv6d/conv_s1/Conv_output_0 tf_output_name: tf.math.add_18/Add:0 shape: (1, 32, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/rebnconv6d/relu_s1/Relu_output_0 tf_output_name: tf.nn.relu_8/Relu:0 shape: (1, 32, 8, 10) dtype: float32 validate_result:  Skipped (Deleted or Shape Unmatched) 
    INFO: onnx_output_name: /stage1/Resize_output_0 tf_output_name: lambda//stage1/Resize:0 shape: (1, 32, 15, 20) dtype: float32 validate_result:  Unmatched  max_abs_error: 1.0628647804260254
    INFO: onnx_output_name: /stage1/Concat_2_output_0 tf_output_name: tf.concat_2/concat:0 shape: (1, 64, 15, 20) dtype: float32 validate_result:  Unmatched  max_abs_error: 1.0628647804260254
    INFO: onnx_output_name: /stage1/rebnconv5d/conv_s1/Conv_output_0 tf_output_name: tf.math.add_20/Add:0 shape: (1, 32, 15, 20) dtype: float32 validate_result:  Unmatched  max_abs_error: 0.6556030511856079
    ```

- onnx
  ![image](https://user-images.githubusercontent.com/33194443/221359900-e4c34413-bcff-49cd-8455-04573a66ed5c.png)
- tflite
  ![image](https://user-images.githubusercontent.com/33194443/221359949-d82ae035-7290-4865-8c7a-fb99c91c8e18.png)

- The atmosphere of normal operation is shown in the figure below.
  https://github.com/PINTO0309/PINTO_model_zoo/tree/main/313_IS-Net

  https://user-images.githubusercontent.com/33194443/221360167-c961c239-19b2-4b10-903f-ec23a26bd7ed.mp4

**Comment by Hyunseok-Kim0** [comment]
This bug is occurs when `ceil_mode` is used. I will make a pull request for the bug fix within this week.

**Comment by Hyunseok-Kim0** [comment]
There is problem in fixing this bug when `count_include_pad` is `False` and `ceil_mode` is `True`.

![Screenshot from 2023-03-03 08-51-41](https://user-images.githubusercontent.com/34959032/222593917-0cf5d42d-5932-4405-be51-e17ff8b0ca15.png)

Left one is output from tflite and right one is from onnx. As shown in the image, nan values exist in onnx output. It looks like if there is no element exists when last stride of pooling, nan value is added due to divided by zero or something when `ceil_mode` is on.
I guess the `ceil_mode` does not treats extra padding same as regular padding. This problem corrupted `average_multiplier`, so the last column and row of tflite has different value for now.

<img src="https://user-images.githubusercontent.com/34959032/222596994-22f24a04-b9de-41ff-a6d6-83eec43c74a7.png" height=75>
This is pytorch output, which have no nan values with different shape. pytorch did not generated third row since there is only padding value in last step.

It is possible to implement to have the same output as onnx. However, the implementation will be like this: zero padding -> pooling -> padding with nan values. This does not make sense and it looks those nan values is kind of bug in onnx itself. I think the implementation should follows pytorch version and report the problem to onnx repo.

**Comment by PINTO0309** [comment]
Thank you.
Hmmm. This is quite a troubling issue. I appreciate your efforts very much.

>I think we should follows pytorch version implementation and report the problem to onnx repo.

I agree.

**Comment by Hyunseok-Kim0** [comment]
Second fix is almost finished but will take couple more days. Also, I reported this problem in onnx repo: https://github.com/onnx/onnx/issues/4958

**Comment by PINTO0309** [comment]
OK. Thanks!

**Comment by PINTO0309** [comment]
Fixes: https://github.com/PINTO0309/onnx2tf/releases/tag/1.7.21
